retry version guard fix

msaroufim opened the pull request.

crcrpar commented:
This does not seem to support source build of pytorch without `PYTORCH_BUILD_VERSION`.
For example, in my local environment, pytorch version is 2.5.0a0+git9f17037 and torchao import is failing as follows:

```
# python -c 'import torch; import torchao'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.10/dist-packages/torchao/__init__.py", line 31, in <module>
    from torchao.quantization import (
  File "/usr/local/lib/python3.10/dist-packages/torchao/quantization/__init__.py", line 7, in <module>
    from .smoothquant import *  # noqa: F403
  File "/usr/local/lib/python3.10/dist-packages/torchao/quantization/smoothquant.py", line 18, in <module>
    from .utils import (
  File "/usr/local/lib/python3.10/dist-packages/torchao/quantization/utils.py", line 11, in <module>
    from torchao.utils import find_multiple
  File "/usr/local/lib/python3.10/dist-packages/torchao/utils.py", line 306, in <module>
    TORCH_VERSION_AT_LEAST_2_5 = torch_version_at_least("2.5.0")
  File "/usr/local/lib/python3.10/dist-packages/torchao/utils.py", line 304, in torch_version_at_least
    return is_fbcode() or compare_versions(torch.__version__, min_version) >= 0
  File "/usr/local/lib/python3.10/dist-packages/torchao/utils.py", line 288, in compare_versions
    v1_parts = parse_version(v1)
  File "/usr/local/lib/python3.10/dist-packages/torchao/utils.py", line 285, in parse_version
    return [int(x) for x in version.split('.')]
  File "/usr/local/lib/python3.10/dist-packages/torchao/utils.py", line 285, in <listcomp>
    return [int(x) for x in version.split('.')]
ValueError: invalid literal for int() with base 10: '0a0'
```

msaroufim commented:
So we've been bit before by having dependencies in general - you can see these 2 threads for more context. The short of it is forcing dependencies was causing forced version downgrades for our partners and most of our dependencies did not really seem necessary
* https://github.com/pytorch/ao/pull/449
* https://github.com/pytorch/ao/issues/371

Unfortunately PyTorch does not depend on packaging https://gist.github.com/msaroufim/09c316d10cd314e189e1b2ad28823c3b but I wish it did

Specifically regarding the version check logic thank you for catching that, this does indeed bother me

I did though try applying your diff and the check is unfortunately also incorrect but if you have a forward fix that would solve this case and the test case I mention the PR description then we should be good. I'll poke around a bit more myself at this

```python
import torch
from torchao.utils import torch_version_at_least

torch.__version__ = "2.5.0a0+git9f17037" # nightly version 

print(torch_version_at_least("2.5.0")) # Returns False
```

``diff
diff --git a/setup.py b/setup.py
index 47f7714..b2ac860 100644
--- a/setup.py
+++ b/setup.py
@@ -131,6 +131,7 @@ setup(
     package_data={
         "torchao.kernel.configs": ["*.pkl"],
     },
+    install_requires=["packaging"],
     ext_modules=get_extensions() if use_cpp != "0" else None,
     extras_require={"dev": read_requirements("dev-requirements.txt")},
     description="Package for applying ao techniques to GPU models",
diff --git a/torchao/utils.py b/torchao/utils.py
index 47227b1..05a04b5 100644
--- a/torchao/utils.py
+++ b/torchao/utils.py
@@ -7,6 +7,8 @@ import torch.nn.utils.parametrize as parametrize
 import itertools
 import time
 import warnings
+from packaging.version import parse
+
 
 __all__ = [
     "benchmark_model",
@@ -279,23 +281,17 @@ def unwrap_tensor_subclass(model, filter_fn=None):
         unwrap_tensor_subclass(child)
     return model
 
-def parse_version(version_string):
-    # Remove any suffixes like '+cu121' or '.dev'
-    version = version_string.split('+')[0].split('.dev')[0]
-    return [int(x) for x in version.split('.')]
 
 def compare_versions(v1, v2):
-    v1_parts = parse_version(v1)
-    v2_parts = parse_version(v2)
-    
-    for i in range(max(len(v1_parts), len(v2_parts))):
-        v1_part = v1_parts[i] if i < len(v1_parts) else 0
-        v2_part = v2_parts[i] if i < len(v2_parts) else 0
-        if v1_part > v2_part:
-            return 1
-        elif v1_part < v2_part:
-            return -1
-    return 0
+    v1_version = parse(v1)
+    v2_version = parse(v2)
+
+    if v1_version == v2_version:
+        return 0
+    elif v1_version > v2_version:
+        return 1
+    else:
+        return -1
 
 def is_fbcode():
     return not hasattr(torch.version, "git_version")

```

ptrblck commented:
NIT:
One shown examples looks wrong:
```python
torch.__version__ = "2.5.0.dev20240708+cu121"
...
print(torch_version_at_least("2.5.0")) # Return True
```
Relase `2.5.0.dev` is not "at least" `2.5.0`, as it's a pre-release (assuming "at least" is interpreted as ">=").
`packaging` also correctly shows is:
```python
import torch
from packaging import version

torch.__version__ = "2.5.0.dev20240708+cu121"

version.parse(torch.__version__) >= version.parse("2.4.0.dev") # True
version.parse(torch.__version__) >= version.parse("2.4.0") # True

version.parse(torch.__version__) >= version.parse("2.5.0.dev") # True
version.parse(torch.__version__) >= version.parse("2.5.0") # False !!!! HERE !!!!

torch.__version__ = "2.5.0"
version.parse(torch.__version__) >= version.parse("2.4.0.dev") # True
version.parse(torch.__version__) >= version.parse("2.4.0") # True

version.parse(torch.__version__) >= version.parse("2.5.0.dev") # True
version.parse(torch.__version__) >= version.parse("2.5.0") # True

torch.__version__ = "2.4.0"
version.parse(torch.__version__) >= version.parse("2.4.0.dev") # True
version.parse(torch.__version__) >= version.parse("2.4.0") # True

version.parse(torch.__version__) >= version.parse("2.5.0.dev") # False
version.parse(torch.__version__) >= version.parse("2.5.0") # False
```

msaroufim commented:
@ptrblck I agree this is somewhat confusing but here's the reasoning https://github.com/pytorch/ao/pull/684#discussion_r1717885424
